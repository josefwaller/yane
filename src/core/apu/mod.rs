mod apu;
pub use apu::Apu;
pub mod envelope;
pub mod length_counter;
mod pulse;
pub use pulse::PulseRegister;
mod triangle;
pub use triangle::TriangleRegister;
mod noise;
pub use noise::NoiseRegister;
mod dmc;
pub use dmc::{DmcRegister, DMC_RATES};
