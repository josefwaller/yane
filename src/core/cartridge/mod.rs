// mod cartridge;
// pub use cartridge::{Cartridge, CartridgeMemory, NametableArrangement};
mod mapper;
pub use mapper::Mapper;
pub mod mappers;

use crate::core::cartridge::mapper::get_mapper;
use log::*;
use serde::{Deserialize, Serialize};
use std::{
    cmp::max,
    fmt::{Debug, Display},
};

#[derive(Debug, PartialEq, Clone, Copy, Serialize, Deserialize)]
/// The various nametable arrangements a cartridge can have.
///
/// Determines how the 2 screens of `VRAM` are mirrored to create the 4 screens of potential outputs.
/// Note that this is the nametable ARRANGEMENT - [NametableArrangement::Horizontal] means that the
/// nametables are using VERTICAL mirroring.
pub enum NametableArrangement {
    OneScreen,
    Horizontal,
    Vertical,
}

/// Contains all memory in the cartridge that isn't mapper-specific.
///
/// Contains PRG/CHR ROM/RAM.
/// Does not contain any latches, banks, or dividers used by mappers.
#[derive(Clone, Serialize, Deserialize)]
pub struct CartridgeMemory {
    /// Program RAM (PRG RAM) of the cartridge
    pub prg_ram: Vec<u8>,
    /// Program ROM (PRG ROM) of the cartridge
    pub prg_rom: Vec<u8>,
    /// Character RAM (CHR RAM) of the cartridge
    pub chr_ram: Vec<u8>,
    /// Character ROM (CHR ROM) of the cartridge
    pub chr_rom: Vec<u8>,
    /// Nametable arrangement of the cartridge, read when parsing the file.
    /// May be changed by the mapper, use [Mapper::nametable_arrangement] to get the current nametable arrangement being used
    pub nametable_arrangement: NametableArrangement,
}
impl CartridgeMemory {
    /// Read a byte from CHR ROM or (if CHR ROM is empty) CHR RAM.
    ///
    /// This is really a convience function, since most cartridges either are all CHR ROM or CHR RAM.
    /// So this could be interpreted as just "Read CHR from whatever format the cartridge using"
    pub fn read_chr(&self, addr: usize) -> u8 {
        if self.chr_rom.is_empty() {
            self.chr_ram[addr % self.chr_ram.len()]
        } else {
            self.chr_rom[addr % self.chr_rom.len()]
        }
    }
    /// Write a byte to CHR RAM, if present.
    pub fn write_chr(&mut self, addr: usize, value: u8) {
        if !self.chr_ram.is_empty() {
            let i = addr % self.chr_ram.len();
            self.chr_ram[i] = value;
        }
    }
}

/// An NES cartridge.
///
/// Contains the cartridge's RAM and ROM in [CartridgeMemory] and a [Mapper] responsible for mapping addresses to data.
#[derive(Serialize, Deserialize)]
pub struct Cartridge {
    /// The memory in the cartridge
    pub memory: CartridgeMemory,
    /// The mapper the cartridge is using
    pub mapper: Box<dyn Mapper>,
    // Whether the cartridge has battery backed RAM and should be saved
    has_battery_ram: bool,
}

impl Cartridge {
    /// Create a new cartridge from the contents of an iNes (.nes) file.
    ///
    /// * `bytes` The contents of the iNes file.
    /// * `savedata` The battery backed static RAM on the cartridge, used to initialise the PRG RAM if present.
    pub fn from_ines(bytes: &[u8], savedata: Option<Vec<u8>>) -> Cartridge {
        if cfg!(debug_assertions) {
            assert_eq!(bytes[0], b'N');
            assert_eq!(bytes[1], b'E');
            assert_eq!(bytes[2], b'S');
            assert_eq!(bytes[3], 0x1A);
        }
        let prg_rom_size = 0x4000 * bytes[4] as usize;
        let chr_rom_size = 0x2000 * bytes[5] as usize;
        let prg_ram_size = max(bytes[8] as usize * 0x2000, 0x2000);
        let mut chr_ram_size = if chr_rom_size == 0 { 0x2000 } else { 0x0 };
        debug!("Cartridge header: {:X?}", &bytes[0..16]);
        let has_battery_ram = (bytes[6] & 0x02) != 0;
        let has_trainer = (bytes[6] & 0x04) != 0;
        let alt_nametable_layout = (bytes[6] & 0x08) != 0;
        debug!(
            "Trainer: {}, alternate nametable: {}, battery backed ram: {}",
            has_trainer, alt_nametable_layout, has_battery_ram
        );
        // Detect type of iNes file
        let total_file_size = 16 + if has_trainer { 512 } else { 0 } + prg_rom_size + chr_rom_size;
        debug!(
            "Total data size: {:X} bytes. File size: {:X}",
            total_file_size,
            bytes.len()
        );
        let file_type = if bytes[7] & 0x0C == 0x08 && bytes.len() >= total_file_size {
            debug!("iNES 2.0 detected");
            0
        } else if bytes[7] & 0x0C == 0x04 {
            debug!("Archaic iNES detected");
            chr_ram_size = 0;
            1
        } else if bytes[7] & 0x0C == 0x00 {
            debug!("iNES detected");
            2
        } else {
            debug!("Archaic iNES probably detected");
            1
        };
        debug!(
            "Detected as {}, ignoring.",
            if bytes[9] & 0x01 != 0 { "PAL" } else { "NTSC" }
        );
        debug!(
            "{:X} bytes PRG ROM, {:X} bytes CHR ROM, {:X} bytes PRG RAM, {:X} bytes CHR RAM",
            prg_rom_size, chr_rom_size, prg_ram_size, chr_ram_size
        );
        // Todo
        let mapper_id = (bytes[6] >> 4) + if file_type != 1 { bytes[7] & 0xF0 } else { 0 };
        let nametable_arrangement = if (bytes[6] & 0x01) == 0 {
            NametableArrangement::Vertical
        } else {
            NametableArrangement::Horizontal
        };
        debug!(
            "Cartridge is using a {:?} nametable arrangment",
            nametable_arrangement
        );
        debug!(
            "Cartridge {} using an alternative nametable arrangement",
            if (bytes[6] & 0x08) == 0 {
                "isn't"
            } else {
                "is"
            }
        );
        debug!(
            "Cartridge is using {} mapper (0x{:X})",
            mapper_id, mapper_id
        );
        let mapper = get_mapper(mapper_id as usize);
        let mut start = 16 + if has_trainer { 512 } else { 0 };
        let mut end = start + prg_rom_size;
        let prg_rom = bytes[start..end].to_vec();
        start = end;
        end += chr_rom_size;
        debug!("Reading CHR ROM at {:#X}", start);
        let chr_rom = bytes[start..end].to_vec();
        // Load PRG RAM from savedata if we have some
        let prg_ram = match savedata {
            Some(data) => {
                assert_eq!(data.len(), prg_ram_size);
                data
            }
            None => vec![0; prg_ram_size],
        };
        Cartridge {
            memory: CartridgeMemory {
                prg_rom,
                chr_rom,
                prg_ram,
                chr_ram: vec![0; chr_ram_size],
                nametable_arrangement,
            },
            mapper,
            has_battery_ram,
        }
    }
    /// Read a byte from the cartridge's memory given an address in CPU memory space
    pub fn read_cpu(&self, addr: usize) -> u8 {
        self.mapper.read_cpu(addr, &self.memory)
    }
    /// Write a byte in the cartridge's memory given an address in CPU memory space
    pub fn write_cpu(&mut self, addr: usize, value: u8) {
        self.mapper.write_cpu(addr, &mut self.memory, value);
    }
    /// Read a byte in the cartridge's memory given an address in PPU memory space
    pub fn read_ppu(&mut self, addr: usize) -> u8 {
        self.mapper.read_ppu(addr, &self.memory)
    }
    /// Write a byte of data to CHR ROM/RAM in PPU memory space.
    pub fn write_ppu(&mut self, addr: usize, value: u8) {
        self.mapper.write_ppu(addr, &mut self.memory, value);
    }
    /// Get all of the CHR data as bytes.
    ///
    /// Only used for debug purposes, the PPU should use [Cartridge::read_ppu] to allow the mapper to transform the address.
    pub fn get_pattern_table(&self) -> &[u8] {
        if self.memory.chr_ram.is_empty() {
            return self.memory.chr_rom.as_slice();
        }
        &self.memory.chr_ram
    }
    /// Transform a given nametable address to a valid address in the PPU's VRAM.
    ///
    /// The NES needs to show four full screens of nametable data (top left, top right, bottom left, and bottom right),
    /// but only has enough memory to store 2 full screens of nametable data.
    /// So two of the screens are mirrored by transforming the addresses when reading nametable data.
    /// See [the NESDEV wiki](https://www.nesdev.org/wiki/PPU_nametables).
    pub fn transform_nametable_addr(&self, addr: usize) -> usize {
        let nametable = self.mapper.nametable_arrangement(&self.memory);
        match nametable {
            NametableArrangement::OneScreen => addr % 0x400,
            NametableArrangement::Horizontal => {
                // 0x2000 = 0x2800, 0x2400 = 0x2C00
                (addr - 0x2000) % 0x800
            }
            NametableArrangement::Vertical => {
                // 0x2000 = 0x2400, 0x2800 = 0x2C00
                if addr < 0x2800 {
                    addr % 0x400
                } else {
                    (addr % 0x400) + 0x400
                }
            }
        }
    }
    /// [true] if the cartridge has battery backed RAM (i.e. save data), [false] otherwise
    pub fn has_battery_backed_ram(&self) -> bool {
        self.has_battery_ram
    }
    /// Get the nametable arrangement the cartridge is currently using
    pub fn nametable_arrangement(&self) -> NametableArrangement {
        self.mapper.nametable_arrangement(&self.memory)
    }
    /// Advance the cartridge by a certain number of CPU cycles
    pub fn advance_cpu_cycles(&mut self, cycles: u32) {
        self.mapper.advance_cpu_cycles(cycles);
    }
}

impl Display for Cartridge {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Display::fmt(&self.mapper, f)
    }
}
impl Debug for Cartridge {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        Debug::fmt(&self.mapper, f)
    }
}
