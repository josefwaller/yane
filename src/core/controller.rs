use serde::{Deserialize, Serialize};
#[cfg(feature = "wasm-bindgen")]
extern crate wasm_bindgen;

/// An NES controller
///
/// Used to represent the controller's state in the emulator.
#[derive(Clone, Copy, Debug, Serialize, Deserialize)]
#[cfg_attr(feature = "wasm-bindgen", wasm_bindgen::prelude::wasm_bindgen)]
pub struct Controller {
    pub up: bool,
    pub left: bool,
    pub right: bool,
    pub down: bool,
    pub start: bool,
    pub select: bool,
    pub a: bool,
    pub b: bool,
}

impl Default for Controller {
    fn default() -> Self {
        Self::new()
    }
}

impl Controller {
    pub fn new() -> Controller {
        Controller {
            up: false,
            left: false,
            right: false,
            down: false,
            start: false,
            select: false,
            a: false,
            b: false,
        }
    }
}
