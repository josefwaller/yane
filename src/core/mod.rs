//! The actual emulation code, provided as a library.
//!
//! A library for emulating the behaviour of the Nintendo Entertainment System.
//! Contains the entire state of the machine, and updates it accordingly as the NES is advanced.
//! Stores the visual output as a 2D array representing the screen, and the audio output
//! as a [VecDeque][std::collections::VecDeque] of samples.
//! ```
//! use yane::core::{Nes, Controller, Settings, HV_TO_RGB};
//! // The actual state of the NES
//! let mut nes = Nes::new();
//! // Various configurable settings for how to run the emulator
//! let settings = Settings::default();
//! // Advance the NES by 1 instruction
//! nes.advance_instruction(&settings);
//! // Advance the NES by 1 frame (continue advancing until a VBlank interval is triggered)
//! nes.advance_frame(&settings);
//! // Press the A button on player 1's controller
//! nes.set_input(0, Controller {
//!   up: false,
//!   left: false,
//!   right: false,
//!   down: false,
//!   a: false,
//!   b: false,
//!   start: false,
//!   select: false
//! });
//! // Read the screen output
//! let screen_output = nes.ppu.output;
//! // Get the RGB value of the top-left pixel
//! let rgb_output = HV_TO_RGB[screen_output[0][0]];
//! println!("Top left pixel is R={} B={} G={}", rgb_output[0], rgb_output[1], rgb_output[2]);
//! // Get the audio output as a vector of samples
//! let audio_output = nes.apu.sample_queue();
//! println!("Read {} audio samples", audio_output.len());
//! ```
mod nes;
pub use nes::{Nes, NesState};
mod cpu;
pub use cpu::Cpu;
mod apu;
pub use apu::Apu;
mod status_register;
pub use status_register::StatusRegister;
mod cartridge;
pub use cartridge::*;
pub mod opcodes;
mod ppu;
pub use ppu::Ppu;
mod controller;
pub use controller::Controller;
mod settings;
pub use settings::Settings;

/// The debug palette, used instead of the palette ram if [Settings::use_debug_palette] is `true`.
pub const DEBUG_PALETTE: [u8; 32] = [
    0x1D, 0x01, 0x11, 0x21, 0x1D, 0x05, 0x15, 0x25, 0x1D, 0x09, 0x19, 0x29, 0x1D, 0x06, 0x16, 0x26,
    0x1D, 0x13, 0x23, 0x33, 0x1D, 0x17, 0x27, 0x37, 0x1D, 0x1B, 0x2B, 0x3B, 0x1D, 0x18, 0x28, 0x38,
];

/// The approximate clock speed of an NES, in hertz.
pub const CPU_CLOCK_SPEED: u32 = 1_789_000;
/// The location of the cartridge's interrupt vector.
pub const CARTRIDGE_IRQ_ADDR: usize = 0xFFFE;
/// The location of the reset interrupt vector.
pub const RESET_IRQ_ADDR: usize = 0xFFFC;
/// The location of the non-maskable interrupt's vector.
pub const NMI_IRQ_ADDR: usize = 0xFFFA;
const PALETTE_DATA: &[u8; 1536] = include_bytes!("../2C02G_wiki.pal");
/// Map of the console's Hue/Value output to RGB values
pub const HV_TO_RGB: &[[u8; 3]] =
    // This is unsafe but it should only be evaluated at compile time
    unsafe { core::slice::from_raw_parts(PALETTE_DATA.as_ptr() as *const [u8; 3], 3 * 64) };
