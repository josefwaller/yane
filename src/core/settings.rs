/// Settings for how to run the emulator.
///
/// Contain fields that change the visual output of the PPU.
/// Some fields can also change the behaviour of some games (by interfering with the
/// sprite 0 hit or sprite overflow flags).
#[derive(Copy, Clone)]
pub struct Settings {
    /// Debugging palette override, assigns each palette a unique colour to quickly show which tiles are using which palettes.
    pub use_debug_palette: bool,
    /// Whether to limit each scanline to rendering at most 8 sprites.
    /// Sprite 0 hit and sprite overflow flag setting behaviour will not be changed, this is only visual
    pub scanline_sprite_limit: bool,
    /// Whether to always draw sprites on top of the background
    pub always_sprites_on_top: bool,
}

impl Default for Settings {
    fn default() -> Self {
        Settings {
            use_debug_palette: false,
            scanline_sprite_limit: false,
            always_sprites_on_top: false,
        }
    }
}
