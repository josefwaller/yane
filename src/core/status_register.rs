use std::fmt::Debug;

use serde::{Deserialize, Serialize};

/// The status register of the NES.

#[derive(Clone, Serialize, Deserialize)]
pub struct StatusRegister {
    /// The carry flag, also known as the unsigned overflow flag
    pub c: bool,
    /// The zero flag
    pub z: bool,
    /// The interrupt diable flag
    pub i: bool,
    /// The decimal mode flag
    pub d: bool,
    /// The break command flag
    pub b: bool,
    /// The (signed) overflow flag
    pub v: bool,
    /// The negative flag
    pub n: bool,
}

impl Default for StatusRegister {
    fn default() -> Self {
        Self::new()
    }
}

impl StatusRegister {
    /// Create a new StatusRegister, initialising all flags to [false].
    pub fn new() -> StatusRegister {
        StatusRegister {
            c: false,
            z: false,
            i: true,
            d: false,
            b: false,
            v: false,
            n: false,
        }
    }
    /// Get the status register as a single byte to be written to memory.
    /// ```
    /// let mut s = yane::core::StatusRegister::new();
    /// s.z = true;
    /// s.d = true;
    /// s.i = false;
    /// assert_eq!(s.to_byte(), 0b00101010);
    /// s.n = true;
    /// s.v = true;
    /// assert_eq!(s.to_byte(), 0b11101010);
    /// ```
    pub fn to_byte(&self) -> u8 {
        let mut b = 0x20;
        if self.c {
            b |= 0x01;
        }
        if self.z {
            b |= 0x02;
        }
        if self.i {
            b |= 0x04;
        }
        if self.d {
            b |= 0x08;
        }
        if self.b {
            b |= 0x10;
        }
        if self.v {
            b |= 0x40;
        }
        if self.n {
            b |= 0x80;
        }
        b
    }
    /// Set the status register from a given byte that contains one bit per flag.
    /// ```
    /// let mut s = yane::core::StatusRegister::new();
    /// s.from_byte(0b11001010);
    /// assert_eq!(s.d, true);
    /// assert_eq!(s.v, true);
    /// assert_eq!(s.d, true);
    /// assert_eq!(s.z, true);
    /// ```
    pub fn from_byte(&mut self, byte: u8) {
        self.c = (byte & 0x01) != 0;
        self.z = (byte & 0x02) != 0;
        self.i = (byte & 0x04) != 0;
        self.d = (byte & 0x08) != 0;
        self.v = (byte & 0x40) != 0;
        self.n = (byte & 0x80) != 0;
    }
}

impl Debug for StatusRegister {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        macro_rules! format_flag {
            ($flag: ident, $name: literal) => {
                if self.$flag {
                    format!(" {}", $name)
                } else {
                    format!("!{}", $name)
                }
            };
        }
        write!(
            f,
            "[{} {} {} {} {} {}]",
            format_flag!(c, "C"),
            format_flag!(z, "Z"),
            format_flag!(i, "I"),
            format_flag!(d, "D"),
            format_flag!(v, "V"),
            format_flag!(n, "N")
        )
    }
}
