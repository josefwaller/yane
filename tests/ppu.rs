use yane::core::{Cartridge, Nes};
mod common;

#[test]
fn test_palette_ram() {
    rom_test!("./test_roms/ppu_palette_ram.nes");
}
#[test]
fn test_oam_ram() {
    rom_test!("./test_roms/ppu_oam_ram.nes");
}
#[test]
fn test_vbl_timing() {
    rom_test!("./test_roms/ppu_vbl_clear_timing.nes");
}
#[test]
fn test_vram_access() {
    rom_test!("./test_roms/ppu_vram_access.nes", 400);
}
#[test]
fn test_open_bus() {
    rom_test!("./test_roms/ppu_open_bus.nes", 300);
}
#[test]
fn test_oam_read() {
    rom_test!("./test_roms/ppu_oam_read.nes");
}
#[test]
fn test_spr_overflow() {
    rom_test!("./test_roms/spr_overflow_basics.nes");
}
#[test]
fn test_spr_overflow_details() {
    rom_test!("./test_roms/spr_overflow_details.nes");
}
