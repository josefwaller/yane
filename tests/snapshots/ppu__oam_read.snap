---
source: tests/ppu.rs
assertion_line: 29
expression: screen.as_slice()
snapshot_kind: text
---
[
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 2D 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 6F 61 6D 5F 72 65 61 64 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 50 61 73 73 65 64 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    "20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20 20",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
    " 0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0  0",
]
